specs/discussions/2025-04-block-proofs.md

```diff
@@ -0,0 +1,319 @@
+     (e.g. `SegmentDescription`)
+   - The root hash of the Mountain Merkle Tree (MMR) of the history of the shard after appending the
+     parent block.
+
+```rust
+struct BlockHeader<T: SegmentDescription>
+	/// Block number
+	number: BlockNumber
+	/// State root (can be considered redundant if tx_root is included)
+	state_root: Hash
+	/// Root of the Merkle tree of transactions included in the block.
+	tx_root: Hash
+	/// Block solution.
+	solution: Proof
+	/// Block number and hash of the beacon block referenced by this hash block.
+	beacon_chain_ref: (BlockNumber, Hash)
+	// The commitment of consensus information for a shard is not stored as
+	// plain transactions in the block, but as their raw data structures (mainly block
+	// and segment information from the child shard).
+	consensus_info_root: Hash
+	/// List of segments and consensus objects that need to be submitted to the upper
+	/// layers of the chain.
+	consensus_info: Vec<T>
+	/// Hash of the parent block.
+	parent_hash: Hash
+	/// Pointer to the MMR root of the history of the shard after appending the parent block.
+	/// This interconnects the root of the history for the child shard after every block,
+	/// allowing to easily generate proofs for the history of the shard.
+	parent_mmr_history_root: Hash
+}
+```
+
+2. Fields in the `BlockHeader` can be arranged in a Merkle Tree to compute the hash of the block as
+   the Merkle proof of this tree instead of directly hashing the whole header. This has an
+   interesting consequences, as it would allow to generate proofs for specific elements of the block
+   header without requiring the whole header.
+
+> Appendix: It is worth structuring the block header as a Merkle tree to generate proofs for
+> specific header fields like `tx_root`, `state_root` `parent_mmr_history_root` and
+> `consensus_info_root` are already represented as root references to the underlying data structured
+> as a tree? A proof for some field in the block header will be of `log_2(num_fields_in_header)` *
+> hash_length`. The savings of hashing the whole header instead of using a tree and having to provide all the fields in the header to verify the block can be computed as `(total_block_header_size-log_2(num_fields_in_header)*hash_length)/total_block_header_size`.
+> If this savings are worth the computational overhead of having to compute the Merkle tree, then we
+> should consider using trees for headers.
+
+3. Farmers submit in a transaction the `BlockHeader` for new blocks to the parent chain. Blocks are
+   only accepted if:
+
+   - References a valid beacon chain block.
+   - The solution is within the right solution range.
+   - It has an increasing block number (unless a fork has happened and can be clearly identified).
+     If the block number of the block being submitted is not the next immediate one and is lower
+     than the latest committed in the parent, it means that a reorg has happened, and the parent
```

Comment:
Which is also a DoS vector, submitting too many reorgs to the parent chain with the goal of overwhelming its capacity. Maybe worth having some minimal delay for block inclusion (like 12 slots or something) during which most of the forks are resolved naturally.